File: IN2090 - Databaser og datamodellering/assignments/IN2090 - Obligatorisk Oppgave 5/administrator.py
"""
                                    marktz
                        IN2090 - Obligatorisk Oppgave 5
                            Programmering med SQL

Forklaring:
Jeg tenkte å gi en liten forklaringen for å gjøre koden mer forståelig.

- O P P G A V E 2 -
Det jeg har gjort i oppgave 2 er først å spørre om username. Hvis brukeren
velger da å ikke skrive noen brukernavn i det hele tatt, vil dette føre til
at strengen blir konvertert til "%%". Dette vil føre til at vi lager "bills"
for alle brukere. Deretter har jeg laget en spørring som henter den nødvendig
informasjon for å løse oppgaven. Til slutt printer jeg resultatet vi får fra
å eksekvere cur.fetchall();

- O P P G A V E 3 -
Jeg har valgt å løse oppgaven på en liten annerledes måte. Jeg synes presentasjonene
til leif var ganske nyttig med tanke på spørrings-delen, men jeg valgte å lage
en egen metode for å hente cid, altså "Kategori"-IDen, når vi setter navnet
på denne. Dersom kategorien ikke finnes, vil vi returnere NONE, som betyr at
navnet på kategorien ikke fantes i databasen, og derfor vil denne ikke ha noen
category-ID.

Konklusjon:
Jeg vet at man kunne ha løst oppgaven på forskjellige måter. Jeg fant denne
løsningen ganske nyttig og lærte masse. Ellers er det bare å komme med
kommentarer til ting som kunne ha blitt løst annerledes.

"""
import psycopg2

# Login details for database user
dbname = "marktz"
user = "marktz_priv"
pwd = "pie0aeCieg" # Set in the password for the _priv-user you got in a mail

# Gather all connection info into one string
connection = \
    "host='dbpg-ifi-kurs01.uio.no' " + \
    "dbname='" + dbname + "' " + \
    "user='" + user + "' " + \
    "port='5432' " + \
    "password='" + pwd + "'"

def admin():
    conn = psycopg2.connect(connection) # Create a connection
    option = True
    while (option == True):
        print("\n--- ADMINISTRATOR --- ")
        print("Please choose an option: \n 1. Create bills \n 2. Insert new product \n 3. Exit")

        user = get_int_from_user("Option: ", True)
        if (user == 1):
            make_bills(conn) # Print bill
        elif (user == 2):
            insert_product(conn) # Insert new product to the database
        elif (user == 3):
            option = False # Ezit program
            print("See you soon!")

"""
Oppgave 2 - Lage regninger
I denne oppgaven skal du implementere funksjonen make_bills (conn) (for python)
som generer regninger for brukere. Funksjonen skal spørre brukeren av programmet
om et brukernavn. Dersom brukeren oppgir et brukernavn skal programmet så
generere en regning for denne brukeren på følgende format:

Name: <navn>
Address: <adresse>
Total due: <total>

hvor <navn> er brukerens navn, <adresse> er brukerens adresse, og <total> er
mengden penger brukeren skylder, altså summen av alle produktene som som
brukeren har bestilt men ennå ikke betalt for (husk at en bestilling kan
inneholde flere av samme produkt via num-kolonnen).
"""

# Oppgave 2
def make_bills(conn):
    print("-- BILLS --")

    # Asking for the username to print the bill
    username = input("Username: ")
    if (username == ""):
        username = "%%"

    query = "SELECT u.name, u.address, SUM(o.num * p.price) " + \
            "FROM ws.orders AS o " + \
                "INNER JOIN ws.products AS p USING (pid) " + \
                "INNER JOIN ws.users AS u USING (uid) " + \
            "WHERE o.payed = 0 AND u.username LIKE %(username)s " + \
            "GROUP  BY u.name, u.address;";

    cur = conn.cursor()
    cur.execute(query, {'username' : username})
    rows = cur.fetchall()

    if (rows == []):
        print("\nNo results.")
        return
    else:
        for row in rows:
            print("\n-- Bill --")
            print("Name: " + row[0] + "\n" + \
                  "Address: " + row[1] + "\n" + \
                  "Total due: " + str(row[2]))

"""
Oppgave 3 - Sette inn nye produkter
I denne oppgaven skal du implementere funksjonen insert_product (conn) som
lar brukeren sette inn nye produkter i Webshopens produktkatalog (som da er
ws.products-tabellen). Brukeren skal bli spurt om navn og pris på produktet,
samt navnet på kategorien produktet tilhører.

-- INSERT NEW PRODUCT -- Product name: Juice
Price: 2.3
Category: food
Description: Fresh orange juice New product Juice inserted.
"""
# Oppgave 3
def insert_product(conn):
    print("-- INSERT NEW PRODUCT --")
    name = input("Product name: ")
    price = input("Price: ")
    while (price == ""):
        price = input("Vennligst tast inn en pris: ")
    price = float(price)
    c = input("Category: ")
    category = hente_cid(conn, c)
    description = input("Description: ")

    query = "INSERT INTO ws.products(name, price, cid, description) " + \
            "VALUES (%(n)s, %(p)s, %(c)s, %(d)s);"

    cur = conn.cursor()
    cur.execute(query, {'n' : name, 'p' : price, 'c' : category, 'd' : description})
    conn.commit()
    print("New product '" + name + "' inserted.")

def hente_cid(conn, category):
    query = "SELECT cid " + \
            "FROM ws.categories " + \
            "WHERE name = %(name)s;"

    cur = conn.cursor()
    cur.execute(query, {'name' : category})
    rows = cur.fetchall()

    if (rows == []):
        return None
    else:
        for row in rows:
            return row[0]

def get_int_from_user(msg, needed):
    # Utility method that gets an int from the user with the first argument as message
    # Second argument is boolean, and if false allows user to not give input,
    # and will then return None
    while True:
        numStr = input(msg)
        if (numStr == "" and not needed):
            return None
        try:
            return int(numStr)
        except:
            print("Please provide an integer or leave blank.")

if __name__ == "__main__":
    admin()

File: IN2090 - Databaser og datamodellering/assignments/IN2090 - Obligatorisk Oppgave 5/Administrator.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Scanner;

public class Administrator{
  public static void main(String[] args){
    String dbname = "marktz"; // Input your UIO-username
    String user = "marktz_priv";   // Input your UIO-username + priv
    String pwd = "pie0aeCieg";    // Input the password for the _priv-user you got in a mail

    String connectionStr =
        "user=" + user + "&" +
        "port=5432&" +
        "password=" + pwd + "";

    String host = "jdbc:postgresql://dbpg-ifi-kurs01.uio.no";
    String connectionURL =
        host + "/" + dbname +
        "?sslmode=require&ssl=true&sslfactory=org.postgresql.ssl.NonValidatingFactory&" +
        connectionStr;

    try {

      // Load driver for PostgreSQL
      Class.forName("org.postgresql.Driver");

      // Create a connectuon to the database
      // Connection connection = DriverManager.getConnection();

      Connection connection = DriverManager.getConnection(host + "/" + dbname
                                                          + "?" + connectionStr);

      boolean program = true;
      while (program == true){
        System.out.println("\n--- ADMINISTRATOR ---");
        System.out.println("Please choose an option: ");
        System.out.println(" 1. Create bills\n 2. Insert new products \n 3. Exit");

        int ch = getIntFromUser("Option: ", true);

        if (ch == 1) {
          makeBills(connection);
        } else if (ch == 2) {
          insertProducts(connection);
        } else if (ch == 3){
          program = false;
          System.out.println("See you soon!");
        }
      }
    } catch (SQLException|ClassNotFoundException ex) {
        System.err.println("Error encountered: " + ex.getMessage());
    }
  }

  private static void makeBills(Connection connection){
    System.out.println("-- BILLS --");

    // Asking for the username to print the bill
    String username = getStrFromUser("Username: ");
    if (username.isEmpty()){
      username = "%%";
    } else {
      username = "" + username + "";
    }

    String query = "SELECT u.name, u.address, SUM(o.num * p.price)" +
                   "FROM ws.orders AS o " +
                      "INNER JOIN ws.products AS p USING (pid) " +
                      "INNER JOIN ws.users AS u USING (uid) " +
                    "WHERE o.payed = 0 AND u.username LIKE ? " +
                    "GROUP BY u.name, u.address;"

    PreparedStatement statement = connection.prepareStatement(query)
    statement.setString(1, username);
    ResultSet rows = statement.executeQuery();

    if (!rows.next()){
      System.out.println("No results.");
      return;
    } else {
      do {
        System.out.println("\n-- Bill --")
        System.out.println("Name: " + rows.getString(1) + "\n" +
                           "Address: " + rows.getString(2) + "\n" +
                           "Total due: " + rows.getInteger(3));
      }
    }

  }

  private static void insertProducts(Connection connection){
    System.out.println("--INSERT NEW PRODUCT");

    String navn = getStrFromUser("Navn paa produkt?", false);
    Int pris = getIntFromUser("Prisen paa produktet?", false);
    String kat = getStrFromUser("Kategori paa produktet?", false);
    String bes = getStrFromUser("Beskrivelse av produktet: ", false);

    String query = "INSERT INTO ws.products as nyP" +
                   "VALUES (c.cid," + navn + "," + pris + "," + kat + "," + bes + ")" +
                   "SELECT c.cid" +
                   "FROM ws.categories as c" +
                   "WHERE name = " + kat;

    prepareStatement statement = connection.prepareStatement(query);
    statement.setString(1, navn);
    statement.setString(2, pris);
    statement.setString(3, kat);
    statement.setString(4, bes);
    ResultSet rows = statement.executeQuery();

    if (!rows.next()) {
      System.out.println("Something went wrong, no results");
    } else {
      do {
        System.out.println("--INSERT NEW PRODUCT");
        System.out.println("Product name: " + rows.getString(1));
        System.out.println("Price: " + rows.getString(2));
        System.out.println("Category: " + rows.getString(3));
        System.out.println("Description: " + rows.getString(4));
        System.out.println("New product " + rows.getString(1) + " inserted.");
      }
    }
  }

  /**
   * Utility method that gets an int as input from user
   * Prints the argument message before getting input
   * If second argument is true, the user does not need to give input and can
   * leave the field blank (resulting in a null)
   */
   private static Integer getIntFromUser(String message, boolean canBeBlank){
     while (true) {
       String str = getStrFromUser(message);
       if (str.equals("") && canBeBlank){
         return null;
       } try {
         return Integer.valueOf(str);
       } catch (NumberFormatException ex){
         System.out.println("Please provide an integer or leave blank.");
       }
     }
   }

   /**
    * Utility method that gets a String as input from user
    * Prints the argument message before getting input
    */
    private static String getStrFromUser(String message){
      Scanner scanner = new Scanner(System.in);
      System.out.println(message);
      return scanner.nextLine();
    }
}
